python/dgl/graphbolt/impl/csc_sampling_graph.py

```diff
@@ -87,15 +88,41 @@ def __init__(
         self._metadata = metadata
 
     @property
-    def num_nodes(self) -> int:
-        """Returns the number of nodes in the graph.
-
+    def num_nodes(self):
+        """
+        >   For a homogenous graph, returns the number of nodes in it.
+        >   For a Heterogenous graph, returns a dictionary indicating
+            the numbers of all types of nodes in it.
         Returns
         -------
         int
-            The number of rows in the dense format.
+            The number of nodes in the entire graph (homogenous).
+        OR
+        Dict[str, int]
+            A dictionary indicating the numbers of all types of nodes
+            (heterogenous).
         """
-        return self._c_csc_graph.num_nodes()
+        offset = self.node_type_offset
+
+        # Homogenous.
+        if offset is None or self.metadata is None:
+            return self._c_csc_graph.num_nodes()
+
+        else:
+            # First, the graph is treated as heterogenous, then the return
+            # value type depends on the length of the dictionary.
+            num_nodes_per_type = {
+                _type: offset[_idx + 1] - offset[_idx]
+                for _type, _idx in self.metadata.node_type_to_id.items()
+            }
+
+            return (
+                num_nodes_per_type
```

Comment:
This logic appears somewhat complex and could potentially be confusing. Could it be simplified for better understanding?